### PR TITLE
[build] Add flags to allow skipping rebuilding the corelibs

### PR DESCRIPTION
**Pull request by finagolfin**
Add three new flags, `--skip-clean-libdispatch`, `--skip-clean-foundation`, and `--skip-clean-xctest`, that leave the previous builds of those products in place.

I've been using a crude version of this patch for years when I'm iterating on one of the later products and don't want the corelibs rebuilt each time.

@edymtt, this is a follow-on to #36686, where you mentioned this is already done for XCTest, but it isn't. Mind reviewing?

### CODE REVIEW

**Comment by buttaface** [comment]
@CodaFi, interested in this pull?

**Comment by buttaface** [comment]
@artemcm, you added some of these other flags in #33563, would you review?

**Review by artemcm — APPROVED** [review]
LGTM!

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by buttaface** [comment]
Oh, I only tested on linux, didn't realize [my expanded test would fail on macOS](https://github.com/apple/swift/blob/8198d82845f84f0c75bd4356fd37f49e7929a33e/utils/build-script-impl#L2464). I will split libdispatch and Foundation off into a linux-only test.

File: utils/build-script-impl [hunk 1]
```diff
@@ -2336,11 +2341,13 @@ for host in "${ALL_HOSTS[@]}"; do
...
-                  # The Swift project might have been changed, but CMake might
-                  # not be aware and will not rebuild.
-                  echo "Cleaning the XCTest build directory"
-                  call rm -rf "${XCTEST_BUILD_DIR}"
+                  if [[ "${SKIP_CLEAN_XCTEST}" == "0" ]]
+                  then
+                    # The Swift project might have been changed, but CMake might
+                    # not be aware and will not rebuild.
+                    echo "Cleaning the XCTest build directory"
+                    call rm -rf "${XCTEST_BUILD_DIR}"
```
Comment (buttaface) [inline utils/build-script-impl:2349]: This is currently only removed for non-Darwin, mistake? I see llbuild is removed by default for all OSs, I can move this above so it's removed for Darwin too, then this new flag will keep it for both.

Comment (buttaface) [reply utils/build-script-impl]: Moved this up so that this build directory is cleaned by default on macOS too, which is now tested by the updated test.

**Comment by buttaface** [comment]
Split the test up and made sure XCTest is cleaned up on macOS too.

File: validation-test/BuildSystem/skip_clean_corelibs.test [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# REQUIRES: standalone_build
+# UNSUPPORTED: OS=macosx
```
Comment (buttaface) [inline validation-test/BuildSystem/skip_clean_corelibs.test:2]: Do I need to mark all Darwin platforms like iOS as unsupported here? I had a previous BuildSystem test pass the CI, then we had to go back and remove all Darwin platforms later, ie #38279.

**Comment by buttaface** [comment]
@drexin, would you run this through the CI again?

**Comment by drexin** [comment]
@swift-ci test

**Comment by buttaface** [comment]
Rebased because #38885 moved a lot of code to a different file and disabled the new test on all Darwin platforms now.

**Comment by buttaface** [comment]
@artemcm, this should pass the CI now.

**Comment by buttaface** [comment]
@CodaFi, would you run this through the CI?

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by buttaface** [comment]
Passed all CI except Windows, whose CI may be stuck.

**Comment by buttaface** [comment]
Ping, passed CI, ready for merge.

**Comment by buttaface** [comment]
@artemcm, anything holding back getting this merged?

**Comment by artemcm** [comment]
@swift-ci please test and merge

**Comment by shahmishal** [comment]
We are seeing failure on CI:
https://ci.swift.org/job/oss-swift-package-macos/6135/console

Changes to `BuildSystem/skip_clean_xctest_llbuild.test` not requires us to checkout `swift-coreslibs-xctest` on macOS platform. This should not be required, can we fix this test to only execute on Linux and Windows?

**Comment by buttaface** [comment]
Huh, that passed the macOS CI though. I will disable the XCTest portion for Darwin platforms, pull forthcoming.

**Pull request merged**